NetKet/Graph/custom_graph.hpp

```diff
@@ -25,13 +25,37 @@
 
 namespace netket {
 
+namespace detail {
+/// Constructs the adjacency list given graph edges. No sanity checks are
+/// performed. Use at your own risk!
+std::vector<std::vector<int>> AdjacencyListFromEdges(
+    const std::vector<AbstractGraph::Edge> &edges, int const number_sites) {
+  assert(number_sites >= 0 && "Bug! Number of sites should be non-negative");
+  std::vector<std::vector<int>> adjacency_list(
+      static_cast<std::size_t>(number_sites));
+  for (auto const &edge : edges) {
+    adjacency_list[edge[0]].push_back(edge[1]);
+    adjacency_list[edge[1]].push_back(edge[0]);
+  }
+  return adjacency_list;
+}
+}  // namespace detail
+
 /**
     Class for user-defined graphs
     The list of edges and nodes is read from a json input file.
 */
 class CustomGraph : public AbstractGraph {
+ public:
+  using AbstractGraph::Edge;
+  using AbstractGraph::ColorMap;
+
+ private:
   // adjacency list
+  // TODO(twesterhout): To be removed
   std::vector<std::vector<int>> adjlist_;
```

Comment:
So you want to work only with edges and get the adjlist only if requested? That is fine to me, in fact the two ways of storing the graph are completely equivalent

Reply:
It's easier to run sanity checks on edges, they require less memory, and (if we want to) we can very easily expose them to Python without copying (I'll show this in the next PR once this one is merged; exposing adjacency lists requires 2D iterators and I'm too lazy to implement one for such a simple task). (Ref. #84)

Reply:
But from the maths perspective, yes, they are equivalent.

Reply:
I am even fine with re-computing adjacency lists when asked by some GetAdjacency method. This is not performance critic, let's stick to simple implementations